core/shared/src/main/scala/zio/ZPool.scala

```diff
@@ -100,94 +117,251 @@ object ZPool {
         case Exit.Success(a) => f(a)
       }
 
-    lazy val toManaged: ZManaged[Any, E, A] = ZIO.done(result).toManaged
+    def toManaged(implicit trace: ZTraceElement): ZManaged[Any, E, A] =
+      ZIO.done(result).toManaged
   }
 
-  private case class DefaultPool[E, A](
+  private case class DefaultPool[R, E, A, S](
     creator: ZManaged[Any, E, A],
     range: Range,
     isShuttingDown: Ref[Boolean],
-    size: Ref[Int],
-    free: Queue[Attempted[E, A]],
-    allocating: Ref[Int],
-    invalidated: Ref[Set[A]]
+    state: Ref[State],
+    items: Queue[Attempted[E, A]],
+    invalidated: Ref[Set[A]],
+    track: Exit[E, A] => UIO[Any]
   ) extends ZPool[E, A] {
 
     /**
-     * Triggers a single allocation in the background. Updates the data
-     * structures to ensure consistency.
+     * Returns the number of items in the pool in excess of the minimum size.
      */
-    private def allocate(implicit trace: ZTraceElement): UIO[Any] =
-      ZIO.unlessZIO(isShuttingDown.get) {
-        ZIO.uninterruptibleMask { restore =>
-          (for {
-            _           <- allocating.update(_ + 1)
-            reservation <- creator.reserve
-            exit        <- restore(reservation.acquire).exit
-            attempted   <- ZIO.succeed(Attempted(exit, reservation.release(Exit.succeed(()))))
-            _           <- free.offer(attempted) *> size.update(_ + 1)
-          } yield attempted).ensuring(allocating.update(_ - 1))
-        }
-      }
+    def excess(implicit trace: ZTraceElement): UIO[Int] =
+      state.get.map { case State(free, size) => size - range.start min free }
```

Comment:
@adamgfraser Why do we need to `min` between `range.start` and `free`? 🤔

It seems to me that `excess` could just be: `size - range.start`, no? 🤔

Reply:
If `free` is smaller, for example because we are in the process of shrinking, then we don't really have that excess.